src/components/CanvasAnimator.tsx

```diff
@@ -1,14 +1,36 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import { Frame, GraphIndex, LabeledGraph } from '../logic/commonInterfaces';
 import Canvas from './Canvas';
 
+const MS_PER_FRAME = 1000
+
 function CanvasAnimator({ graph, frames }
   : {
     graph: LabeledGraph,
     frames: Array<Frame>,
   }) {
+  const [animation, setAnimation] = useState({
```

Comment:
Storing an object as a state might be not the best idea. In general, it's better to use immutable data types for the state. Here you can easily decompose the object. (I can't find the docs now but I'm pretty sure React team suggests decomposing the state in that manner)

Reply:
What what I've seen it is normal practice. Even react [documentation]() uses arrays and objects. From my experience operating with nested objects might be a bit tricky, but also doable.

Reply:
Please show me the paragraph where they store objects in functional components. Why u prefer to do something "a bit tricky"? 

Also, which one do you think is easier to understand?

`setAnimation({ ...animation, frameNum: nextFrame });`

`setFramNumber(nextFrame)`

Reply:
I don't prefer using nested objects. That is why this is single-layer object. Rest of the animation object is useful and it wouldn't be wise to just throw it away. Solutions:
 - A: Store it in single object
 - B: use multiple `useState`
[Here](https://reactjs.org/docs/hooks-state.html#tip-using-multiple-state-variables) is very complex `useState` (First frame, last line).

Reply:
That's why I suggest going B bc it's compliant to the React conventions. (Your example is the array - obviously there is no way to decompose it)

Reply:
Alright, [here](https://reactjs.org/docs/hooks-faq.html#should-i-use-one-or-many-state-variables) it both solutions are okay as long as it looks nice. I'll go with B